File: src/java/non-corba/com/metricstream/omg/naming/NamingContextImpl.java
package com.metricstream.omg.naming;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Implementation of the NamingContext interface.
 * This is the non-CORBA equivalent of the CORBA NamingContext implementation.
 */
public class NamingContextImpl implements NamingContext {
    
    /**
     * Map of bindings in this context, keyed by the string representation of the name component.
     */
    private final Map<String, BindingEntry> bindings = new ConcurrentHashMap<>();
    
    /**
     * Entry in the bindings map, containing the object and its binding type.
     */
    private static class BindingEntry {
        Object object;
        BindingType type;
        
        BindingEntry(Object object, BindingType type) {
            this.object = object;
            this.type = type;
        }
    }
    
    /**
     * Creates a new naming context implementation.
     */
    public NamingContextImpl() {
        // Nothing to initialize
    }
    
    @Override
    public void bind(Name n, Object obj) throws NotFound, CannotProceed, InvalidName, AlreadyBound {
        if (n == null || obj == null) {
            throw new InvalidName("Name or object is null");
        }
        
        if (n.size() == 0) {
            throw new InvalidName("Empty name");
        }
        
        if (n.size() == 1) {
            // Binding directly in this context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            synchronized (bindings) {
                if (bindings.containsKey(key)) {
                    throw new AlreadyBound("Name already bound: " + key);
                }
                
                BindingType type = (obj instanceof NamingContext) ? BindingType.ncontext : BindingType.nobject;
                bindings.put(key, new BindingEntry(obj, type));
            }
        } else {
            // Binding in a sub-context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            BindingEntry entry = bindings.get(key);
            if (entry == null) {
                throw new NotFound(NotFoundReason.missing_node, n);
            }
            
            if (entry.type != BindingType.ncontext) {
                throw new NotFound(NotFoundReason.not_context, n);
            }
            
            NamingContext subContext = (NamingContext) entry.object;
            subContext.bind(n.suffix(1), obj);
        }
    }
    
    @Override
    public void rebind(Name n, Object obj) throws NotFound, CannotProceed, InvalidName {
        if (n == null || obj == null) {
            throw new InvalidName("Name or object is null");
        }
        
        if (n.size() == 0) {
            throw new InvalidName("Empty name");
        }
        
        if (n.size() == 1) {
            // Rebinding directly in this context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            BindingType type = (obj instanceof NamingContext) ? BindingType.ncontext : BindingType.nobject;
            bindings.put(key, new BindingEntry(obj, type));
        } else {
            // Rebinding in a sub-context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            BindingEntry entry = bindings.get(key);
            if (entry == null) {
                throw new NotFound(NotFoundReason.missing_node, n);
            }
            
            if (entry.type != BindingType.ncontext) {
                throw new NotFound(NotFoundReason.not_context, n);
            }
            
            NamingContext subContext = (NamingContext) entry.object;
            subContext.rebind(n.suffix(1), obj);
        }
    }
    
    @Override
    public NamingContext bind_new_context(Name n) throws NotFound, CannotProceed, InvalidName, AlreadyBound {
        if (n == null) {
            throw new InvalidName("Name is null");
        }
        
        if (n.size() == 0) {
            throw new InvalidName("Empty name");
        }
        
        if (n.size() == 1) {
            // Creating a new context directly in this context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            synchronized (bindings) {
                if (bindings.containsKey(key)) {
                    throw new AlreadyBound("Name already bound: " + key);
                }
                
                NamingContext newContext = new NamingContextImpl();
                bindings.put(key, new BindingEntry(newContext, BindingType.ncontext));
                return newContext;
            }
        } else {
            // Creating a new context in a sub-context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            BindingEntry entry = bindings.get(key);
            if (entry == null) {
                throw new NotFound(NotFoundReason.missing_node, n);
            }
            
            if (entry.type != BindingType.ncontext) {
                throw new NotFound(NotFoundReason.not_context, n);
            }
            
            NamingContext subContext = (NamingContext) entry.object;
            return subContext.bind_new_context(n.suffix(1));
        }
    }
    
    @Override
    public Object resolve(Name n) throws NotFound, CannotProceed, InvalidName {
        if (n == null) {
            throw new InvalidName("Name is null");
        }
        
        if (n.size() == 0) {
            throw new InvalidName("Empty name");
        }
        
        if (n.size() == 1) {
            // Resolving directly in this context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            BindingEntry entry = bindings.get(key);
            if (entry == null) {
                throw new NotFound(NotFoundReason.missing_node, n);
            }
            
            return entry.object;
        } else {
            // Resolving in a sub-context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            BindingEntry entry = bindings.get(key);
            if (entry == null) {
                throw new NotFound(NotFoundReason.missing_node, n);
            }
            
            if (entry.type != BindingType.ncontext) {
                throw new NotFound(NotFoundReason.not_context, n);
            }
            
            NamingContext subContext = (NamingContext) entry.object;
            return subContext.resolve(n.suffix(1));
        }
    }
    
    @Override
    public void unbind(Name n) throws NotFound, CannotProceed, InvalidName {
        if (n == null) {
            throw new InvalidName("Name is null");
        }
        
        if (n.size() == 0) {
            throw new InvalidName("Empty name");
        }
        
        if (n.size() == 1) {
            // Unbinding directly in this context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            BindingEntry entry = bindings.remove(key);
            if (entry == null) {
                throw new NotFound(NotFoundReason.missing_node, n);
            }
        } else {
            // Unbinding in a sub-context
            NameComponent nc = n.get(0);
            String key = getKey(nc);
            
            BindingEntry entry = bindings.get(key);
            if (entry == null) {
                throw new NotFound(NotFoundReason.missing_node, n);
            }
            
            if (entry.type != BindingType.ncontext) {
                throw new NotFound(NotFoundReason.not_context, n);
            }
            
            NamingContext subContext = (NamingContext) entry.object;
            subContext.unbind(n.suffix(1));
        }
    }
    
    @Override
    public void list(int how_many, BindingListHolder bl, BindingIteratorHolder bi) {
        if (bl == null) {
            bl = new BindingListHolder(new BindingList());
        } else if (bl.value == null) {
            bl.value = new BindingList();
        }
        
        List<Binding> allBindings = new ArrayList<>();
        
        for (Map.Entry<String, BindingEntry> entry : bindings.entrySet()) {
            String key = entry.getKey();
            BindingEntry bindingEntry = entry.getValue();
            
            NameComponent nc = parseKey(key);
            Name name = new Name();
            name.addComponent(nc);
            
            Binding binding = new Binding(name, bindingEntry.type);
            allBindings.add(binding);
        }
        
        // Fill the binding list with up to how_many bindings
        int count = Math.min(how_many, allBindings.size());
        for (int i = 0; i < count; i++) {
            bl.value.add(allBindings.get(i));
        }
        
        // Create an iterator for the remaining bindings if necessary
        if (count < allBindings.size() && bi != null) {
            List<Binding> remainingBindings = allBindings.subList(count, allBindings.size());
            bi.value = new BindingIteratorImpl(remainingBindings);
        } else if (bi != null) {
            bi.value = null;
        }
    }
    
    @Override
    public void destroy() throws NotEmpty {
        if (!bindings.isEmpty()) {
            throw new NotEmpty("Cannot destroy non-empty naming context");
        }
    }
    
    /**
     * Gets a key for the bindings map from a name component.
     * 
     * @param nc The name component
     * @return The key
     */
    private String getKey(NameComponent nc) {
        if (nc.kind == null || nc.kind.isEmpty()) {
            return nc.id;
        } else {
            return nc.id + "." + nc.kind;
        }
    }
    
    /**
     * Parses a key from the bindings map into a name component.
     * 
     * @param key The key
     * @return The name component
     */
    private NameComponent parseKey(String key) {
        int dotIndex = key.indexOf('.');
        if (dotIndex >= 0) {
            String id = key.substring(0, dotIndex);
            String kind = key.substring(dotIndex + 1);
            return new NameComponent(id, kind);
        } else {
            return new NameComponent(key, "");
        }
    }
}

File: src/main/kotlin/com/metricstream/naming/NamingContextImpl.kt
package com.metricstream.naming

import mu.KotlinLogging
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean

private val logger = KotlinLogging.logger {}

/**
 * Implementation of the NamingContext interface.
 * Provides a thread-safe naming context using ConcurrentHashMap.
 */
class NamingContextImpl : NamingContext {
    private val bindings = ConcurrentHashMap<String, Pair<Any, BindingType>>()
    private val destroyed = AtomicBoolean(false)
    
    /**
     * Checks if this context has been destroyed.
     * @throws CannotProceed if the context has been destroyed
     */
    private fun checkDestroyed() {
        if (destroyed.get()) {
            throw CannotProceed(this, Name())
        }
    }
    
    /**
     * Validates that the name is not empty.
     * @param name The name to validate
     * @throws InvalidName if the name is empty
     */
    private fun validateName(name: Name) {
        if (name.size() == 0) {
            throw InvalidName()
        }
    }
    
    /**
     * Creates a key for the bindings map from a NameComponent.
     * @param component The name component
     * @return The key string
     */
    private fun createKey(component: NameComponent): String {
        return "${component.id}|${component.kind}"
    }
    
    override fun bind(name: Name, obj: Any) {
        checkDestroyed()
        validateName(name)
        
        if (name.size() == 1) {
            // Binding directly in this context
            val key = createKey(name.components[0])
            if (bindings.containsKey(key)) {
                throw AlreadyBound()
            }
            bindings[key] = obj to BindingType.OBJECT
            logger.debug { "Bound object to name: $name" }
        } else {
            // Need to traverse to the right context
            val firstComponent = name.components[0]
            val key = createKey(firstComponent)
            val pair = bindings[key] ?: throw NotFound(NotFoundReason.MISSING_NODE, name)
            
            if (pair.second != BindingType.CONTEXT) {
                throw NotFound(NotFoundReason.NOT_CONTEXT, name)
            }
            
            val context = pair.first as? NamingContext 
                ?: throw NotFound(NotFoundReason.NOT_CONTEXT, name)
            
            try {
                context.bind(name.suffix(1), obj)
            } catch (e: NotFound) {
                throw NotFound(e.reason, Name((listOf(firstComponent) + e.restOfName.components).toMutableList()))
            } catch (e: CannotProceed) {
                throw CannotProceed(this, name)
            }
        }
    }
    
    override fun bind_context(name: Name, context: NamingContext) {
        checkDestroyed()
        validateName(name)
        
        if (name.size() == 1) {
            // Binding directly in this context
            val key = createKey(name.components[0])
            if (bindings.containsKey(key)) {
                throw AlreadyBound()
            }
            bindings[key] = context to BindingType.CONTEXT
            logger.debug { "Bound context to name: $name" }
        } else {
            // Need to traverse to the right context
            val firstComponent = name.components[0]
            val key = createKey(firstComponent)
            val pair = bindings[key] ?: throw NotFound(NotFoundReason.MISSING_NODE, name)
            
            if (pair.second != BindingType.CONTEXT) {
                throw NotFound(NotFoundReason.NOT_CONTEXT, name)
            }
            
            val existingContext = pair.first as? NamingContext 
                ?: throw NotFound(NotFoundReason.NOT_CONTEXT, name)
            
            try {
                existingContext.bind_context(name.suffix(1), context)
            } catch (e: NotFound) {
                throw NotFound(e.reason, Name((listOf(firstComponent) + e.restOfName.components).toMutableList()))
            } catch (e: CannotProceed) {
                throw CannotProceed(this, name)
            }
        }
    }
    
    override fun rebind(name: Name, obj: Any) {
        checkDestroyed()
        validateName(name)
        
        try {
            unbind(name)
        } catch (e: NotFound) {
            // Ignore, we'll create it
        }
        
        try {
            bind(name, obj)
        } catch (e: AlreadyBound) {
            // This shouldn't happen as we just unbound it
            logger.error { "Unexpected AlreadyBound exception after unbind: $e" }
            throw CannotProceed(this, name)
        }
    }
    
    override fun rebind_context(name: Name, context: NamingContext) {
        checkDestroyed()
        validateName(name)
        
        try {
            unbind(name)
        } catch (e: NotFound) {
            // Ignore, we'll create it
        }
        
        try {
            bind_context(name, context)
        } catch (e: AlreadyBound) {
            // This shouldn't happen as we just unbound it
            logger.error { "Unexpected AlreadyBound exception after unbind: $e" }
            throw CannotProceed(this, name)
        }
    }
    
    override fun resolve(name: Name): Any {
        checkDestroyed()
        validateName(name)
        
        if (name.size() == 1) {
            // Resolving directly in this context
            val key = createKey(name.components[0])
            val pair = bindings[key] ?: throw NotFound(NotFoundReason.MISSING_NODE, name)
            return pair.first
        } else {
            // Need to traverse to the right context
            val firstComponent = name.components[0]
            val key = createKey(firstComponent)
            val pair = bindings[key] ?: throw NotFound(NotFoundReason.MISSING_NODE, name)
            
            if (pair.second != BindingType.CONTEXT) {
                throw NotFound(NotFoundReason.NOT_CONTEXT, name)
            }
            
            val context = pair.first as? NamingContext 
                ?: throw NotFound(NotFoundReason.NOT_CONTEXT, name)
            
            try {
                return context.resolve(name.suffix(1))
            } catch (e: NotFound) {
                throw NotFound(e.reason, Name((listOf(firstComponent) + e.restOfName.components).toMutableList()))
            } catch (e: CannotProceed) {
                throw CannotProceed(this, name)
            }
        }
    }
    
    override fun unbind(name: Name) {
        checkDestroyed()
        validateName(name)
        
        if (name.size() == 1) {
            // Unbinding directly in this context
            val key = createKey(name.components[0])
            if (!bindings.containsKey(key)) {
                throw NotFound(NotFoundReason.MISSING_NODE, name)
            }
            bindings.remove(key)
            logger.debug { "Unbound name: $name" }
        } else {
            // Need to traverse to the right context
            val firstComponent = name.components[0]
            val key = createKey(firstComponent)
            val pair = bindings[key] ?: throw NotFound(NotFoundReason.MISSING_NODE, name)
            
            if (pair.second != BindingType.CONTEXT) {
                throw NotFound(NotFoundReason.NOT_CONTEXT, name)
            }
            
            val context = pair.first as? NamingContext 
                ?: throw NotFound(NotFoundReason.NOT_CONTEXT, name)
            
            try {
                context.unbind(name.suffix(1))
            } catch (e: NotFound) {
                throw NotFound(e.reason, Name((listOf(firstComponent) + e.restOfName.components).toMutableList()))
            } catch (e: CannotProceed) {
                throw CannotProceed(this, name)
            }
        }
    }
    
    override fun bind_new_context(name: Name): NamingContext {
        checkDestroyed()
        validateName(name)
        
        val newContext = NamingContextImpl()
        bind_context(name, newContext)
        return newContext
    }
    
    override fun destroy() {
        checkDestroyed()
        
        if (!bindings.isEmpty()) {
            throw NotEmpty()
        }
        
        destroyed.set(true)
        logger.debug { "Context destroyed" }
    }
    
    override fun list(howMany: Int, bindingList: BindingList): BindingIterator? {
        checkDestroyed()
        
        val allBindings = bindings.map { (key, pair) ->
            val parts = key.split("|", limit = 2)
            val nameComponent = NameComponent(parts[0], parts[1])
            Binding(Name(mutableListOf(nameComponent)), pair.second)
        }
        
        val iterator = BindingIteratorImpl(allBindings)
        
        // Fill the binding list with the first howMany bindings
        iterator.next_n(howMany, bindingList)
        
        // Return the iterator only if there are more bindings
        return if (iterator.hasMore()) iterator else null
    }
}

/**
 * Implementation of the BindingIterator interface.
 */
class BindingIteratorImpl(private val bindings: List<Binding>) : BindingIterator {
    private var index = 0
    private val destroyed = AtomicBoolean(false)
    
    /**
     * Checks if this iterator has been destroyed.
     */
    private fun checkDestroyed() {
        if (destroyed.get()) {
            throw IllegalStateException("Iterator has been destroyed")
        }
    }
    
    /**
     * Checks if there are more bindings.
     * @return true if there are more bindings
     */
    fun hasMore(): Boolean {
        return index < bindings.size
    }
    
    override fun next_one(): Binding? {
        checkDestroyed()
        
        if (!hasMore()) {
            return null
        }
        
        return bindings[index++]
    }
    
    override fun next_n(howMany: Int, bindingList: BindingList): Boolean {
        checkDestroyed()
        
        var count = 0
        while (count < howMany && hasMore()) {
            next_one()?.let {
                bindingList.add(it)
                count++
            }
        }
        
        return hasMore()
    }
    
    override fun destroy() {
        destroyed.set(true)
    }
}
